cas/src/main/java/org/springframework/security/cas/web/CasAuthenticationFilter.java

```diff
@@ -226,6 +236,29 @@ protected final void successfulAuthentication(HttpServletRequest request, HttpSe
 		chain.doFilter(request, response);
 	}
 
+	/**
+	 * We override this method because we don't want to clear the rememberMe in case of an
+	 * unsuccessful CAS gateway request.
+	 */
+	@Override
+	protected void unsuccessfulAuthentication(HttpServletRequest request, HttpServletResponse response,
+			AuthenticationException failed) throws IOException, ServletException {
+
+		// if the request had a non empty artifact, we really encounter an unsuccessful
+		// authentication
+		// else it is probably an CAS gateway request with no SSO session, we redirect to
+		// the saved url
+		if (StringUtils.hasText(obtainArtifact(request))) {
+			super.unsuccessfulAuthentication(request, response, failed);
+		}
+		else {
+			SavedRequest savedRequest = this.requestCache.getRequest(request, response);
+			if (savedRequest != null) {
+				this.redirectStrategy.sendRedirect(request, response, savedRequest.getRedirectUrl());
+			}
+		}
```

Comment:
I'm finding this strategy to decide if that was a gateway request too sensitive.

Is it possible to make CAS redirect to the service URL with a query param? Something like `my.company.com/cas/login?gateway=true&service=https://server3.company.com/webapp/login/cas%3Fgateway=true`. With that, upon successful/failed authentication, the CAS server would redirect to `https://server3.company.com/webapp/login/cas?gateway=true` and we can be sure whether that was a gateway request or not.

Reply:
You're right, this strategy is a bit too sensitive (checking if we already have a saved request).

Though, using a parameter in the `serviceUrl` will trigger a lot of complexity. Indeed, the service tickets created by the CAS server are linked to the original service URL.
It means that we can't reuse the existing `ticketValidator` to validate service tickets created on a gateway request as the service URL won't match, it has `gateway=true` for example while we try to validate the service ticket for the service without `gateway=true`.

I can give it a try if you want. I just want you to confirm.

Reply:
Let's discard the parameter for now. What about setting some attribute in the session (if the session exists) and checking that attribute later on

Reply:
By the way, sorry for the back and forth on this, I haven't had time to test how this works on my own yet.

Reply:
OK. I will try to use the session for that.

Reply:
No problem. Good design takes time ;-)

Reply:
In fact, removing the exception has a drawback: without exception, there is no saved request so I need to do that manually in the `TriggerCasGatewayFilter`

Reply:
No problem in that case, the `OAuth2AuthorizationRequestRedirectFilter` also saves the request before the redirect.

Reply:
OK. Good

Reply:
@marcusdacoregio I have updated the PR to take into account all your feedbacks:
- the exception has been removed
- the `TriggerCasGatewayFilter` performs the redirection to the CAS server with `gateway=true`
- the `CasAuthenticationEntryPoint` is no longer changed
- the `gateway` nature of the CAS round-trip is saved into the web session.

BTW, I have added a `TriggerCasGatewayFilterTests` to test the `TriggerCasGatewayFilter` and cropped all commits into one.

We should be good for merging. Just let me know.

Reply:
@marcusdacoregio Did you get some time to check this PR?

Reply:
I'll take a look at it today